### PR TITLE
web: make web integration tests great again!

### PR DESCRIPTION
**Pull request by valerybugakov**
**TL;DR**: shave off 50% of the time from the Pupetteer build step.
**DISCLAIMER** Diff in the PR is quite small. `yarn.lock` update adds most of the lines changed 🙂

## Context

Our Puppeteer integration tests are slow. This PR relies on [Percy's ability to parallelize visual testing](https://docs.percy.io/docs/parallel-test-suites) on multiple Builkite agents. How it works:

1. Puppeteer preparation agent builds web application and uploads an artifact to be shared between agents.
2. We find all web integration test files, split them in chunks by 3 (empirically found number), and start a separate Buildkite agent for each chunk. 
3. Chunk agent downloads pre-built artifact and runs integrations tests on a small subset of test files.
4. When all tests are completed, the final Buildkite agent kicks in to finalize Percy build.

## Changes

- Split web integration tests in chunks to leverage Percy's parallelization.
- [Percy CLI](https://github.com/percy/cli) is updated to the latest version.
- Chromium version used by Puppeteer is updated to one of the latest versions.
- Added missing methods to the Buildkite golang package to add dependent steps to the pipeline.

## Results

1. **~8 mins** to execute all integration tests.
2. This can be scaled _indefinitely_ with more integration tests being added to the web application.

Screenshot 2021-09-21 at 15 34 26

### Pipeline generated for integration tests

Screenshot 2021-09-21 at 15 34 41

Related work from @jhchabran https://github.com/sourcegraph/sourcegraph/pull/25158.

### CODE REVIEW

File: enterprise/dev/ci/internal/ci/pipeline-steps.go [hunk 1]
```diff
@@ -92,33 +95,72 @@ func addBrowserExt(pipeline *bk.Pipeline) {
 		bk.Cmd("dev/ci/codecov.sh -c -F typescript -F unit"))
 }
 
+func addClientIntegrationTests(pipeline *bk.Pipeline) {
+	CHUNK_SIZE := 3
+	PREP_STEP_KEY := "puppeteer:prep"
+	SKIP_GIT_CLONE_STEP := bk.Plugin("uber-workflow/run-without-clone", "")
```
Comment (valerybugakov) [inline enterprise/dev/ci/internal/ci/pipeline-steps.go:101]: What's the right naming convention to use here?

File: client/shared/src/testing/integration/polly/CdpAdapter.ts [hunk 2]
```diff
@@ -324,7 +324,14 @@ export class CdpAdapter extends PollyAdapter {
         event: Protocol.Fetch.RequestPausedEvent,
         cdpSession: Puppeteer.CDPSession
     ): void {
-        const { requestId } = event
+        const { requestId, request } = event
+
+        // Passthrough missing Chrome extension request because it's expected to fail if the extension is not installed.
+        if (request.url.includes('chrome-extension://invalid')) {
```
Comment (valerybugakov) [inline client/shared/src/testing/integration/polly/CdpAdapter.ts:330]: This is now required because CDP pauses requests to `chrome-extension://invalid`, which was not the case on the earlier version of the Chromium that Puppeteer used for tests.

File: client/web/src/integration/utils.ts [hunk 3]
```diff
@@ -102,7 +102,7 @@ export const percySnapshotWithVariants = async (
 
     try {
         // Wait for Monaco editor to finish rendering before taking screenshots
-        await page.waitForSelector('#monaco-query-input .monaco-editor', { visible: true })
+        await page.waitForSelector('#monaco-query-input .monaco-editor', { visible: true, timeout: 1000 })
```
Comment (valerybugakov) [inline client/web/src/integration/utils.ts:105]: That's a nice win. In a lot of cases, we waited for 30 seconds until the next because there's no `#monaco-query-input` on the page.

File: enterprise/dev/ci/internal/ci/pipeline-steps.go (hunk 1 above)
Comment (jhchabran) [reply enterprise/dev/ci/internal/ci/pipeline-steps.go:101]:
```suggestion
	chunkSize := 3
	prepStepKey := "puppeteer:prep"
	skipGitCloneStep := bk.Plugin("uber-workflow/run-without-clone", "")
```

Comment (valerybugakov) [reply enterprise/dev/ci/internal/ci/pipeline-steps.go:101]: Thanks!

File: enterprise/dev/ci/internal/ci/pipeline-steps.go [hunk 4]
```diff
@@ -64,13 +64,16 @@ func addWebApp(pipeline *bk.Pipeline) {
 		bk.Cmd("dev/ci/codecov.sh -c -F typescript -F unit"))
 }
 
+// Don't download browser, we use "download-puppeteer-browser" script instead
```
Comment (umpox) [inline enterprise/dev/ci/internal/ci/pipeline-steps.go:67]:
```suggestion
// We provide our own Chromium instance that is installed through the `download-puppeteer-browser` script
```

This reads a bit better now that we've changed the code to no longer reference skip download, or the script

File: enterprise/dev/ci/internal/ci/pipeline-steps.go [hunk 5]
```diff
@@ -92,33 +95,72 @@ func addBrowserExt(pipeline *bk.Pipeline) {
...
+	chunkSize := 3
+	prepStepKey := "puppeteer:prep"
+	skipGitCloneStep := bk.Plugin("uber-workflow/run-without-clone", "")
+
+	// Build web application used for integration tests to share it between multiple parallel steps.
+	pipeline.AddStep(":puppeteer::electric_plug: Puppeteer tests prep",
+		bk.Key(prepStepKey),
+		bk.Env("ENTERPRISE", "1"),
+		bk.Cmd("COVERAGE_INSTRUMENT=true dev/ci/yarn-build.sh client/web"),
+		bk.Cmd("dev/ci/create-client-artifact.sh"))
```
Comment (umpox) [inline enterprise/dev/ci/internal/ci/pipeline-steps.go:108]: Nice!

**Review by umpox — APPROVED** [review]
This is so good 👏

File: enterprise/dev/ci/internal/ci/pipeline-steps.go [hunk 6]
```diff
@@ -92,33 +95,72 @@ func addBrowserExt(pipeline *bk.Pipeline) {
...
+		stepKey := fmt.Sprintf("puppeteer:chunk:%s", fmt.Sprint(i+1))
+		puppeteerFinalizeDependencies[i] = bk.DependsOn(stepKey)
+
+		pipeline.AddStep(stepLabel,
+			skipGitCloneStep,
+			bk.Key(stepKey),
+			bk.DependsOn(prepStepKey),
+			percyBrowserExecutableEnv,
+			bk.Env("PERCY_ON", "true"),
+			bk.Cmd(fmt.Sprintf(`dev/ci/yarn-web-integration.sh "%s"`, chunkTestFiles)),
```
Comment (jhchabran) [inline enterprise/dev/ci/internal/ci/pipeline-steps.go:128]: It seems that there is some sort of sync issue here, the two failing steps [here](https://buildkite.com/sourcegraph/sourcegraph/builds/108935#07be853a-3d51-44f7-9ebb-50fc19d117c1) are run at 17:49:37 and the green ones are ran at 17:49:38.

File: client/shared/src/testing/puppeteer-browser-revision.ts [hunk 7]
```diff
@@ -2,7 +2,7 @@
  * The browser revision used by Puppeteer, which is downloaded by BrowserFetcher.
  */
 export const PUPPETEER_BROWSER_REVISION: { chrome: string; firefox: string } = {
-    chrome: '818858',
+    chrome: '885174',
```
Comment (vovakulikov) [inline client/shared/src/testing/puppeteer-browser-revision.ts:5]: Just out of curiosity why exactly this revision? I thought it's the latest but as I see here https://github.com/puppeteer/puppeteer/blob/main/src/revisions.ts#L23 the latest one is 901912 or so.

**Review by vovakulikov — APPROVED** [review]

File: client/shared/src/testing/puppeteer-browser-revision.ts (hunk 7 above)
Comment (valerybugakov) [reply client/shared/src/testing/puppeteer-browser-revision.ts:5]: I relied on the revision that Puppeteer downloads if no executable path is provided and then found the one available for both Mac and Linux platforms. For some reason, it wasn't a `90****` revision at the moment of testing. Happy to try the version that you found in the Puppeteer repo.  Thanks for looking into it!

File: enterprise/dev/ci/internal/ci/pipeline-steps.go (hunk 6 above)
Comment (valerybugakov) [reply enterprise/dev/ci/internal/ci/pipeline-steps.go:127]: Trying to disable the skip-git-clone plugin.

File: client/shared/src/testing/puppeteer-browser-revision.ts (hunk 7 above)
Comment (valerybugakov) [reply client/shared/src/testing/puppeteer-browser-revision.ts]: Updated to 901812

**Review by bobheadxi — APPROVED** [review]
Awesome work!
